### PR TITLE
Worldcat init & logging init

### PR DESCRIPTION
**Pull request by klinga**
Hi Mike,
This is a bit messy and larger than usual pull. Sorry about that. I've been tweaking constantly things that has already been reviewed. bot.py and manager.py are mostly a placeholders at this point as I was trying to figure how how the app should work. Please excuse commented out code, although maybe this will give you clues how my thinking goes.

Per your advise I began creation of some basic structure:
+ comms : Worldcat, BPL Solr, NYPL Platform methods
+ marc : parsing, manipulating, writing MARC records
+ config: logging, etc.

I've tried to use some typing, but I did not apply it consistently across the board yet. I will clean it up in the next pull.

### CODE REVIEW

File: nightshift/comms/worldcat.py [hunk 1]
```diff
@@ -0,0 +1,228 @@
+# -*- coding: utf-8 -*-
+
+"""
+This module handles WorldCat Metadata API requests.
+"""
```
Comment (mwbenowitz) [inline nightshift/comms/worldcat.py:5]: I think it would make sense to convert this to a class. The main advantage I think is that it would enable you to generate an access token when you instantiate the class (probably in `manager.py`?) instead of re-generating it. That same structure could be used to re-use a session object, which might be easier and give you the option to catch/handle session errors if the token or session expire.

File: nightshift/comms/worldcat.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+    Returns:
+        bool
+    """
+    if response.json()["numberOfRecords"] == 0:
+        return False
+    else:
+        return True
+
+
+def prep_resource_queries_payloads(resource: Resource) -> List[dict]:
```
Comment (mwbenowitz) [inline nightshift/comms/worldcat.py]: I think this could be simplified by using a look-up dict for the values. That is just personal preference because I find the long if/elif chains hard to follow. (I'm still not convinced that leaving out `switch` from python was the right call haha)

File: nightshift/comms/worldcat.py [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                logger.debug(
+                    f"Full bib request for resource.nid={resource.nid} : {response.url}"
+                )
+                yield (resource, response)
+            except WorldcatSessionError:
+                logger.error("WorldcatSessionError. Aborting.")
+                raise
+
+
+def get_oclc_number(response: Response) -> str:
```
Comment (mwbenowitz) [inline nightshift/comms/worldcat.py]: Does it make sense to sub-class or otherwise wrap `Response` in a custom class that can contain some of this logic? It feels like there are two things happening in this module: 1) Loading multiple records from WorldCat and 2) transforming them into a more usable format. I think dividing those responsibilities might make the code easier to parse. (this is another personal preference one)

File: nightshift/datastore.py [hunk 4]
```diff
@@ -229,16 +233,14 @@ class WorldcatQuery(Base):
 
     nid = Column(Integer, primary_key=True)
     resourceId = Column(Integer, ForeignKey("resource.nid"), nullable=False)
-    libraryId = Column(Integer, ForeignKey("library.nid"), nullable=False)
     match = Column(Boolean, nullable=False)
-    responseCode = Column(Integer)
-    response = Column(PickleType)  # save as requests.Response object?
+    response = Column(JSON)
```
Comment (mwbenowitz) [inline nightshift/datastore.py]: It might be better to store this as `JSONB` as that gives better options for querying the data (if that's something that might be done in the future).

File: nightshift/datastore_transactions.py [hunk 5]
```diff
@@ -54,12 +64,117 @@ def insert_or_ignore(session, model, **kwargs):
         return None
 
 
+def retrieve_full_bib_resources(session: Session, libraryId: int) -> List[Resource]:
```
Comment (mwbenowitz) [inline nightshift/datastore_transactions.py]: What is the difference between this and `retrieve_matched_resources`? They seem to largely overlap if not entirely overlap

File: nightshift/manager.py [hunk 6]
```diff
@@ -0,0 +1,143 @@
+"""
+This module incldues top level processes to be performed by the app
+"""
```
Comment (mwbenowitz) [inline nightshift/manager.py:3]: I would also make this module a class, I think it could help make the relationship between these methods clearer.

Also could you explain the difference in fetching the brief vs full bibs from worldcat? I think I'm getting a bit lost in the process of fully fleshing out the local catalog records.

**Review by mwbenowitz — APPROVED** [review]
I left some comments but overall this looks really good! I'm able to follow the flow of logic in all the modules and I think the reorganization makes sense.

I think it will be easier to follow once the rest of the code is in place, but it would also be nice if there is a clear explanation or a chart in the Readme with the specific process here. I get the idea but it would be good to feel confident that I fully understand the process.

**Comment by mwbenowitz** [comment]
Also on the coveralls failure, I don't think a `99.7%` coverage is really something that needs to be addressed haha

File: nightshift/datastore_transactions.py (hunk 5 above)
Comment (klinga) [reply nightshift/datastore_transactions.py]: @mwbenowitz 
Matched resources are those with identified matching Worldcat record, but not yet obtained full bib (MARC XML).
Full bib resources are those with downloaded MARC XML that can be upgraded/enhanced. Maybe better naming would clear this up.
I opted for processing batches of records at each stage: first searching, then downloading full bibs, finally producing MARC records for these that were matched to a Worldcat bib. This flow seems to force tagging resources so they are picked up if needed in the next process.

There is some duplication between Resource.status and oclcMatchNumber and fullBib columns. I think searches for appropriate resources could rely on presence or absence of values in two last columns instead. I will try to think about it again and reorganize it.

File: nightshift/comms/worldcat.py (hunk 1 above)
Comment (klinga) [reply nightshift/comms/worldcat.py:5]: That makes sense. I will try to refactor this module. 
`MetadataSession` has build-in handling of expired token. It checks it before each request is issued and obtains new one if needed.

File: nightshift/comms/worldcat.py (hunk 2 above)
Comment (klinga) [reply nightshift/comms/worldcat.py]: That's a great suggestion.
It looks like 3.10 introduces something similar to `switch`:
https://docs.python.org/3.10/tutorial/controlflow.html#match-statements

File: nightshift/datastore.py (hunk 4 above)
Comment (klinga) [reply nightshift/datastore.py]: Thanks!

File: nightshift/comms/worldcat.py (hunk 2 above)
Comment (mwbenowitz) [reply nightshift/comms/worldcat.py]: oh wow they are! That's wild, people have been talking about that for 20+ years (https://www.python.org/dev/peps/pep-3103/)

File: nightshift/comms/worldcat.py (hunk 1 above)
Comment (mwbenowitz) [reply nightshift/comms/worldcat.py:5]: okay cool, that simplifies things a lot!

Comment (klinga) [reply nightshift/comms/worldcat.py:5]: Replaced the code in this module with two new classes `BriefBibResponse` and `Worldcat`.
As mentioned refreshing tokens is handled by bookops-worldcat, but I suspect when we get to processing actual records in the wild we will encounter some issues with OCLC service.

File: nightshift/comms/worldcat.py (hunk 2 above)
Comment (klinga) [reply nightshift/comms/worldcat.py]: @mwbenowitz 
I've been having problems creating a simple lookup dictionary here. `q` parameter varies from resource to resource based on the category of material and available identifiers found on the bib in Sierra.  I failed to find other way to pass these values and construct appropriate payloads. Any advise?

File: nightshift/comms/worldcat.py (hunk 3 above)
Comment (klinga) [reply nightshift/comms/worldcat.py]: It makes a lot of sense. I added `BriefBibResponse` class to handle some of the logic related to parsing/transformation of the responses.

File: nightshift/manager.py (hunk 6 above)
Comment (klinga) [reply nightshift/manager.py:3]: Oops, I missed your comment. I'll have to think about how to do that. I will try in the next PR.

The new OCLC MetadataAPI search endpoints return only brief bibs in JSON. In the past a flag could be passed to the service to return full bib. That is no longer supported in the new version of their API. To obtain full bib if oclc number is unknown, one needs to make two requests: first search and find oclc # of the best matching record in Worldcat and then make another request to a different endpoint to request full bib as MARCXML. 
search: https://developer.api.oclc.org/wc-metadata-v1-1
full bibs: https://developer.api.oclc.org/wc-metadata

**Comment by klinga** [comment]
@mwbenowitz 
If that would not be a trouble would you be able to review my changes? The code differs quite a bit and maybe you will spot some better solutions. TIA!
